runtime/state.go

```diff
@@ -0,0 +1,117 @@
+package runtime
+
+import (
+	log "github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	kerror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8 "k8s.io/client-go/kubernetes"
+)
+
+// State handles the state for service bundles
+type State struct {
+	Client k8.Interface
+	// this is the namespace where the master configs will be stored
+	targetNs string
+}
+
+// NewState returns a state manager
+func NewState(client k8.Interface, targetNamespace string) *State {
+	s := &State{
+		Client:   client,
+		targetNs: targetNamespace,
+	}
+	return s
+}
+
+// CopyState copies the state configmap from one namespace to another
+func (s *State) CopyState(fromName, toName, fromNS, toNS string) error {
+	log.Info("state: copying state from namespace " + fromNS + " to ns " + toNS + " from name " + fromName + " to name " + toName)
+	fromClient := s.Client.CoreV1().ConfigMaps(fromNS)
+	toClient := s.Client.CoreV1().ConfigMaps(toNS)
+	fromMap, err := fromClient.Get(fromName, metav1.GetOptions{})
+	if err != nil {
+		if kerror.IsNotFound(err) {
+			//can't copy if there is nothing to copy
+			log.Info("no state configmap to copy")
+			return nil
+		}
+		return err
+	}
+	toMap, err := toClient.Get(toName, metav1.GetOptions{})
+	if err != nil {
+		if kerror.IsNotFound(err) {
+			fromMap.Namespace = toNS
+			fromMap.Name = toName
+			fromMap.ResourceVersion = ""
+			if _, err := toClient.Create(fromMap); err != nil {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+	for k, v := range fromMap.Data {
+		toMap.Data[k] = v
+	}
+	if _, err := toClient.Update(toMap); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Name provides a consistent name for the state object
+func (s *State) Name(id string) string {
+	return id + "-state"
+}
+
+// StateIsPresent checks to see is there an object carrying state for ServiceBundle
+func (s *State) StateIsPresent(stateName string) (bool, error) {
+	log.Info("state: checking is present")
+	if _, err := s.Client.CoreV1().ConfigMaps(s.targetNs).Get(stateName, metav1.GetOptions{}); err != nil {
+		if kerror.IsNotFound(err) {
+			log.Info("state: is not found ** ")
```

Comment:
How often will this occur? Feels like this should be a debug statement not info.